File: framework/api/src/capability/kfx-control.ts
// Public narrow capability for the KFX Control Suite. Product roles identify
// assembly/distribution metadata only. Core owns Passport verification,
// capability grants, Work/Warrant settlement, CAS, and recovery state.

import type { KfxHost } from './kfx-host.js';
import type { KfLocator, KfNativeBinding } from './types.js';
import { resolveRuntimeDir } from './types.js';

export type KfxControlRoot = {
  kind: 'product';
  path: string;
};

export type KfxControlAuthority = Record<string, unknown>;

export type KfxControlStatus = {
  schema: 'kungfu.kfx.control-suite-status/v1';
  controllerId: 'kungfu-kfx-control-suite';
  statusRoot: string;
  cutRoot: string | null;
  revision: number;
  mode: 'active' | 'safe-mode';
  executionAllowed: boolean;
  active: null | {
    packageRoot: string;
    manifestRoot: string;
    version: string;
  };
  lastKnownGood: null | {
    packageRoot: string;
    manifestRoot: string;
    version: string;
    sourcePath: string;
  };
  diagnostics: Array<{
    code: 'KF_KFX_CONTROL_SAFE_MODE';
    recoveryGuidance: string[];
  }>;
};

export type KfxControlPlan = {
  schema: 'kungfu.kfx.control-suite-plan/v1';
  controllerId: 'kungfu-kfx-control-suite';
  operation: 'install' | 'update';
  packageKey: 'kfx-manager';
  controlPlanRoot: string;
  bootstrapPolicyRoot: string;
  authorizationPlanRoot: string;
  capabilityGrantRoot: string;
  warrantRoot: string;
  allowed: true;
  requiresAuthorization: true;
  candidate: {
    packageRoot: string;
    manifestRoot: string;
    version: string;
  };
  loadPlan: {
    cutRoot: string | null;
    revision: number;
    registryRoot: string;
    graphRoot: string;
    planRoot: string;
    packages: Array<{
      key: string;
      packageRoot: string;
      trustRoot: string;
    }>;
  };
};

export type KfxControlApplication = {
  schema: 'kungfu.kfx.control-suite-application/v1';
  controllerId: 'kungfu-kfx-control-suite';
  controlPlanRoot: string;
  bootstrapPolicyRoot: string;
  authorizationPlanRoot: string;
  capabilityGrantRoot: string;
  warrantRoot: string;
  verified: true;
  status: KfxControlStatus;
  application: Record<string, unknown>;
};

export type KfxControlHostProjection = {
  schema: 'kungfu.kfx.control-host-projection/v1';
  host: KfxHost;
  controllerId: 'kungfu-kfx-control-suite';
  statusRoot: string;
  cutRoot: string | null;
  revision: number;
  mode: 'active' | 'safe-mode';
  executionAllowed: boolean;
  diagnostics: KfxControlStatus['diagnostics'];
};

export function projectKfxControlSuiteHost(
  status: KfxControlStatus,
  host: KfxHost,
): KfxControlHostProjection {
  if (
    status.schema !== 'kungfu.kfx.control-suite-status/v1' ||
    status.controllerId !== 'kungfu-kfx-control-suite' ||
    !status.statusRoot.startsWith('sha256:') ||
    (status.mode === 'active') !== status.executionAllowed ||
    (status.cutRoot === null) !== (status.revision === 0)
  ) {
    throw new Error('KFX Control status identity does not match');
  }
  return {
    schema: 'kungfu.kfx.control-host-projection/v1',
    host,
    controllerId: status.controllerId,
    statusRoot: status.statusRoot,
    cutRoot: status.cutRoot,
    revision: status.revision,
    mode: status.mode,
    executionAllowed: status.executionAllowed,
    diagnostics: status.diagnostics,
  };
}

export type KfxControl = {
  status: () => KfxControlStatus;
  plan: (
    operation: 'install' | 'update',
    candidate: KfxControlRoot,
    authority: KfxControlAuthority,
  ) => KfxControlPlan;
  apply: (
    operation: 'install' | 'update',
    candidate: KfxControlRoot,
    plan: KfxControlPlan,
    authority: KfxControlAuthority,
    auditActor: string,
  ) => KfxControlApplication;
};

export type OpenKfxControlOptions = {
  binding: KfNativeBinding;
  locator: KfLocator;
};

export function openKfxControl(options: OpenKfxControlOptions): KfxControl {
  const operation = options.binding.runStorageServiceOperation;
  if (!operation) {
    throw new Error('native binding does not expose KFX Control operations');
  }
  const runtimeDir = resolveRuntimeDir(options.locator);
  const run = <T>(action: 'status' | 'plan' | 'apply', request: object): T =>
    operation('kfx_runtime', runtimeDir, { action, request }) as T;
  const requestFor = (
    action: 'install' | 'update',
    candidate: KfxControlRoot,
    authority: KfxControlAuthority,
  ) => ({
    ...authority,
    controller: 'kungfu-kfx-control-suite',
    packageKey: 'kfx-manager',
    operation: action,
    roots: [candidate],
  });
  return {
    status: () =>
      run<KfxControlStatus>('status', {
        controller: 'kungfu-kfx-control-suite',
      }),
    plan: (action, candidate, authority) =>
      run<KfxControlPlan>('plan', requestFor(action, candidate, authority)),
    apply: (action, candidate, plan, authority, auditActor) => {
      if (!auditActor.trim()) {
        throw new Error('KFX Control audit actor is required');
      }
      const member = plan.loadPlan.packages.find(
        (row) => row.key === 'kfx-manager',
      );
      if (!member) {
        throw new Error('KFX Control plan does not contain kfx-manager');
      }
      return run<KfxControlApplication>('apply', {
        ...requestFor(action, candidate, authority),
        expectedCutRoot: plan.loadPlan.cutRoot,
        expectedRevision: plan.loadPlan.revision,
        expectedRegistryRoot: plan.loadPlan.registryRoot,
        expectedGraphRoot: plan.loadPlan.graphRoot,
        expectedPlanRoot: plan.loadPlan.planRoot,
        expectedTrustRoot: member.trustRoot,
        expectedPackageRoot: member.packageRoot,
        expectedControlPlanRoot: plan.controlPlanRoot,
        expectedBootstrapPolicyRoot: plan.bootstrapPolicyRoot,
        expectedAuthorizationPlanRoot: plan.authorizationPlanRoot,
        expectedCapabilityGrantRoot: plan.capabilityGrantRoot,
        expectedWarrantRoot: plan.warrantRoot,
        actor: auditActor,
      });
    },
  };
}

File: framework/core/src/python/kungfu/kfx_control.py
"""Thin CLI/Agent adapter over the Core-owned KFX Control Suite protocol."""

from __future__ import annotations

from pathlib import Path
from typing import Any

from kungfu.storage import service as storage_service


CONTROLLER_ID = "kungfu-kfx-control-suite"
PACKAGE_KEY = "kfx-manager"


def _request(candidate: str | Path, operation: str) -> dict[str, Any]:
    if operation not in {"install", "update"}:
        raise ValueError(f"unsupported KFX Control operation: {operation}")
    return {
        "controller": CONTROLLER_ID,
        "packageKey": PACKAGE_KEY,
        "operation": operation,
        "roots": [{"kind": "product", "path": str(Path(candidate).resolve())}],
        "runtimeTiers": {PACKAGE_KEY: "first-party-pinned"},
    }


def status(runtime_dir: str | Path) -> dict[str, Any]:
    return storage_service.kfx_registry(
        "status", {"controller": CONTROLLER_ID}, runtime_dir
    )


def plan(
    runtime_dir: str | Path, candidate: str | Path, operation: str
) -> dict[str, Any]:
    return storage_service.kfx_registry(
        "plan", _request(candidate, operation), runtime_dir
    )


def apply(
    runtime_dir: str | Path,
    candidate: str | Path,
    control_plan: dict[str, Any],
    authorized_by: str,
) -> dict[str, Any]:
    if not authorized_by.strip():
        raise ValueError("KFX Control authorization identity is required")
    load_plan = control_plan["loadPlan"]
    package = next(
        row for row in load_plan["packages"] if row.get("key") == PACKAGE_KEY
    )
    request = {
        **_request(candidate, str(control_plan["operation"])),
        "expectedCutRoot": load_plan["cutRoot"],
        "expectedRevision": load_plan["revision"],
        "expectedRegistryRoot": load_plan["registryRoot"],
        "expectedGraphRoot": load_plan["graphRoot"],
        "expectedPlanRoot": load_plan["planRoot"],
        "expectedTrustRoot": package["trustRoot"],
        "expectedPackageRoot": package["packageRoot"],
        "expectedControlPlanRoot": control_plan["controlPlanRoot"],
        "expectedBootstrapPolicyRoot": control_plan["bootstrapPolicyRoot"],
        "authorizationId": authorized_by.strip(),
        "actor": authorized_by.strip(),
    }
    return storage_service.kfx_registry("apply", request, runtime_dir)
